### PR TITLE
Fix "standard", part 1

### PR DESCRIPTION
**Pull request by zetok**

### CODE REVIEW

File: message_persistence/message_persistence.md [hunk 1]
```diff
@@ -1,10 +1,13 @@
 Message Persistence
 ===================
 
-- ![](/badge/rec.png) A client should provide an option to save its message
+- ![](/badge/req.png) A client should provide an option to save its message
```
Comment (JFreegman) [inline message_persistence/message_persistence.md:4]: There's no reason for this to be required.

File: data_storage/export_format.md [hunk 2]
```diff
@@ -1,14 +1,16 @@
...
 
-- ![](/badge/req.png) If the client supports user avatars, they will be stored
-  under the `avatars/` directory in the root of the client's working directory.
+- ![](/badge/rec.png) If the client supports user avatars, they should be stored
+  under the `avatars/` directory in the root of directory selected by the user.
 
 - ![](/badge/req.png) Avatars are named according to the following format:
   `PK.png`, where PK is the 64-character hexadecimal public key of its owner.
+  Additionally, aside from avatar data, hash of an avatarshould be stored in the
+  following format `PK.hash`.
```
Comment (JFreegman) [inline data_storage/export_format.md:13]: There's no reason to require saving hashes, as calculating them on the fly is trivial.

File: data_storage/export_format.md [hunk 3]
```diff
@@ -1,14 +1,16 @@
...
 
-- ![](/badge/req.png) A Toxcore data file saved in the root of the client's
-  working directory, with the file name `<profile_name>.tox`
+- ![](/badge/req.png) A Toxcore data file saved in the directory selected by
+  the user, with the file name `<profile_name>.tox`
 
-- ![](/badge/req.png) If the client supports user avatars, they will be stored
-  under the `avatars/` directory in the root of the client's working directory.
+- ![](/badge/rec.png) If the client supports user avatars, they should be stored
+  under the `avatars/` directory in the root of directory selected by the user.
```
Comment (JFreegman) [inline data_storage/export_format.md:8]: Explanation?

File: data_storage/export_format.md [hunk 4]
```diff
@@ -1,14 +1,16 @@
 Export Format
 -------------
 
-- ![](/badge/req.png) A Toxcore data file saved in the root of the client's
-  working directory, with the file name `<profile_name>.tox`
+- ![](/badge/req.png) A Toxcore data file saved in the directory selected by
+  the user, with the file name `<profile_name>.tox`
```
Comment (JFreegman) [inline data_storage/export_format.md:5]: Explanation?

File: message_persistence/message_persistence.md (hunk 1 above)
Comment (zetok) [inline message_persistence/message_persistence.md:4]: UX is the reason.

File: data_storage/export_format.md (hunk 4 above)
Comment (zetok) [inline data_storage/export_format.md:5]: UX.

File: data_storage/export_format.md (hunk 3 above)
Comment (zetok) [inline data_storage/export_format.md:8]: UX.

File: data_storage/export_format.md (hunk 2 above)
Comment (zetok) [inline data_storage/export_format.md:13]: > The client must check the file ID against a previously-saved hash of the friend's avatar. If it is the same, the client must cancel the transfer, and use the previously-stored avatar.

File: data_storage/export_format.md (hunk 4 above)
Comment (JFreegman) [inline data_storage/export_format.md:5]: UX is not the main concern of the standard. The focus is on client compatibility and security, as the introduction says.

Comment (zetok) [inline data_storage/export_format.md:5]: Unless you're forgetting about something, "security" is ~useless without UX.

File: message_persistence/message_persistence.md (hunk 1 above)
Comment (stal888) [inline message_persistence/message_persistence.md:4]: The DTS allows developers to choose what kind of UX they want.

File: data_storage/export_format.md (hunk 4 above)
Comment (stal888) [inline data_storage/export_format.md:5]: There is a reason this document is not called the "qTox User Experience Standard".

**Comment by JFreegman** [comment]
It's worth mentioning here that this new standard is focused on enforcing best security practices and client interoperability, particularly related to behaviour that toxcore itself can't enforce. This is stated in the introduction. 

Clauses that are solely related to style and user experience should be strongly justified, and should never be required. We don't intend to position ourselves as style/UX arbiters; we want to encourage our client developers to be as creative as they like, so long as they meet certain minimum requirements.

File: data_storage/export_format.md (hunk 2 above)
Comment (JFreegman) [inline data_storage/export_format.md]: Yep my bad, ignore that comment.

Comment (JFreegman) [inline data_storage/export_format.md:13]: It should say

> The client must check the file hash against the hash of the friend's currently saved avatar.

**Comment by zetok** [comment]
> Clauses that are solely related to style and user experience should be strongly justified

Where is a justification of export format? (and preferably other things too)

AFAICS it seemed like worst part of this standard - I was planning to make issue about it, but since the topic of justification was touched here.. `:)`

> We don't intend to position ourselves as style/UX arbiters;

That's interesting claim - how exactly you do that by requiring clients to put data in specified, hardcoded path that user has no knowledge about? (this question isn't only about that, but also relates to other points across standard)

**Comment by JFreegman** [comment]
> Where is a justification of export format? (and preferably other things too)
> 
> how exactly you do that by requiring clients to put data in specified, hardcoded path that user has no knowledge about?

These are both related to client compatibility, which is our focus. Also, we are working on a rationale that goes along with the standard itself.

**Comment by zetok** [comment]
> These are both related to client compatibility, which is our focus.

I see.

The problem though is that `export format` has unencrypted data because of "compatibility" - this is a ~huge problem.

IMHO something like https://github.com/tux3/qTox/issues/1656 should be done - having it in standard would help with compatibility, and getting encryption would ensure better security.

# 

Though.

> > how exactly you do that by requiring clients to put data in specified, hardcoded path that user has no knowledge about?
> 
> These are both related to client compatibility, which is our focus.

**Storing something in working dirdoesn't help with compatibility**.

Unless you think about "working dir" only in context of ~Linux - then though your point about "compatibility" doesn't make a lot of sense…

Now, by requiring clients to make user decide where that data should be stored, you're improving both security, and compatibility, given that user can save data where they want, and then point client to the place where they saved data.

File: message_persistence/message_persistence.md (hunk 1 above)
Comment (zetok) [inline message_persistence/message_persistence.md:4]: Well then, if UX isn't the main focus, and this standard distances self from it, I guess I can revert this change - just want to confirm - you want that, right?

**Comment by stal888** [comment]
You are misunderstanding the purpose of the export format. The first point under data storage:

> A client is allowed to store profile data internally in any format.

We don't care what clients do with data until it is being moved between clients, at that point, it becomes a compatibility issue in the TCS's scope.

The "client's working directory" is defined as wherever the profile is saved. This could be a client saving internally in ~/.config or as an explicit user export to /Users/tr/Desktop/my_tox_profile.

**Comment by Zer0-One** [comment]
I probably shouldn't have used "working directory" here. I'm going to change that to something that very clearly specifies only "the directory where a given client stores its profile data".

File: message_persistence/message_persistence.md (hunk 1 above)
Comment (ovalseven8) [inline message_persistence/message_persistence.md:4]: 'Recommendation' is fine, I think.

**Comment by Chuongv** [comment]
Hey @zetok , this PR has some  merge conflicts. Could you break up this PR and resubmit? It will be a lot easier for us to review each one individually and bring it to a vote. Thanks :smile:

**Comment by Zer0-One** [comment]
Given that this has gone a month without response, I'm going to close it. If @zetok decides to get back to us, we can re-open this.

**Pull request closed without merging**